### ISSUE TITLE
rapidly repeated DNS query

### ISSUE THREAD

**Issue by jhodal** [turbot/steampipe-plugin-aws#1177]
**Describe the bug**
Ran the below query for cloudwatch log streams last week.  After a few moments, received an error stating my DNS server is "misbehaving."  I'm not clear if that error text is from Steampipe, an AWS API response, or something else.

As it turns out, my own local DNS server was the source of the timeout, there were 1000 DNS queries made from my laptop to the DNS server in less than 60 seconds.  If I'm interpreting this correctly, my steampipe query resulted in ~1000 consecutive repeated lookups for `logs.us-east-1.amazonaws.com` before my DNS server (rightfully) cut me off.  For a few minutes, my laptop is completely cut-off from being able to query DNS, at all, and I lose general connectivity.  I'm basically banned from using DNS for hammering it a thousand consecutive times with the same query.

**Steampipe version (`steampipe -v`)**
steampipe (mac) v0.15.0
aws plugin v0.66.0

**To reproduce**
`steampipe query 'select region,first_event_timestamp,last_event_timestamp,log_group_name,name from my_connection_name.aws_cloudwatch_log_stream' --output csv > log-streams.csv`

**Expected behavior**
a csv file from "my_connection_name.aws_cloudwatch_log_stream" containing the fields from my select statement

**Additional context**
- I was able to get around this by adding logs.us-east-1.amazonaws.com to my hosts file so no queries go to my DNS server, but I don't feel like this behavior of rapidly repeating the same DNS query is efficient.
- querying a smaller AWS account with less streams seems to be successful.

Including a screenshot of my console output as well as my local eth & wifi ip addresses getting temporarily banned on my home pi-hole DNS server.
![image](https://user-images.githubusercontent.com/35462152/179606795-5520a68b-f4aa-429f-b6ee-4355c809cb0c.png)

**Comment by jhodal** [comment]
> Hey @jhodal , in the connection you're testing, which regions have you included, is it just us-east-1? Also, approximately how many log streams do you anticipate you have across all of those regions?
> 
> In terms of number of requests, in the `aws_cloudwatch_log_stream` table, we request up to 50 log streams per request (https://github.com/turbot/steampipe-plugin-aws/blob/main/aws/table_aws_cloudwatch_log_stream.go#L118-L120), which is the maximum allowed by the AWS API. So in terms number of DNS requests, I wonder if we're potentially looking at 50k+ number of log streams?

Hi @cbruno10 thanks for investigating with me.  According to my aws.spc file, the following regions would be queried:

- us-east-1
- us-east-2
- us-west-2

I would suspect 99% of the streams would be found in us-east-1.

As for your other question, i don't know how many log streams there would potentially be; ironically that's one of the questions I'm trying to answer before we start cleaning this up.  But to indirectly answer your question, I have an incomplete CSV output (probably used my modified hosts file to get farther) and I'm seeing 84k lines in that file.  So it looks more than likely I'm hitting limits on the AWS API side once I circumvent the DNS bottleneck.  Well over the 50k+ records you were hinting at.

For what it's worth, i'm not overly concerned about getting the complete list at this point.  In a roundabout way, the answer to how many streams we have turns out to be "way too many streams."  But I am curious about the nature of the rapid/repeat DNS query.

Anyway, please let me know if there is any other info I can provide.  Also, this is a reproducible behavior.

**Comment by cbruno10** [comment]
Thanks @jhodal for the information! As a workaround for now, are there any patterns to your log groups that you can use to break up some of the queries? For instance, instead of running something like `select * from aws_cloudwatch_log_group`, could you run the following queries sequentially:
- `select * from aws_cloudwatch_log_group where name_prefix = '/aws/lambda/';`
- `select * from aws_cloudwatch_log_group where name_prefix = '/aws/macie/';`
- `select * from aws_cloudwatch_log_group where name_prefix = 'OtherPrefix';`

This may help to slow down the outgoing DNS requests and prevent the server from flagging potential abuse.

Please note that in the currently available AWS plugin version, there's a bug in the `aws_cloudwatch_log_group` table that prevents querying on the name prefix correctly, so I've staged a fix in https://github.com/turbot/steampipe-plugin-aws/tree/update-cw-log-group-prefix-qual. If you'd like to try this branch out, please clone the repo and run `make` inside of the cloned repo to locally build a plugin version.

**Comment by cbruno10** [comment]
@kaidaguerre Based on the data provided by @jhodal , the number of requests seems to be expected (84k/50 = 1680+ requests). In the AWS Go SDK (v1 and v2), I couldn't find any way to limit the number of connections or slow down how many requests are allowed in a given timeframe.

Previously we had thrown around ideas of limiting how many connections Steampipe can create at once, though I'm not sure that would help here since this table is not doing any hydrate calls, but instead just calling the `DescribeLogGroups` API page after page.

Is there anything available in the Steampipe SDK that I'm missing that could help in this scenario? Or is this something we should instead handle in the AWS plugin code specifically?

**Comment by e-gineer** [comment]
@cbruno10 What has me curious is why this creates so many DNS requests specifically? I understand why it would be a lot of AWS requests, but I don't understand why the same DNS name would need to be resolved 1000+ times.

That DNS resolution would be from the AWS plugin down I think - any ideas or search results on the DNS resolution in particular?

**Comment by cbruno10** [comment]
@e-gineer Good question, I checked the table code and the code that creates the CloudWatch Logs service client as a starting point, both look as expected and I verified the service client is getting cached correctly. Initial searches across the AWS SDK v1 and v2 issues haven't turned up anything yet about potential DNS resolution issues, but I'll keep searching and see if I can find anything in the issues or code.

**Comment by jhodal** [comment]
@cbruno10 , I re-ran the query as per your suggestion, with a filter (for %lambda%) the query gets interrupted from aws <top half of screenshot>.  The outputted csv file is 1 byte.  So it's unclear if the filter makes a difference.  it simply looks like AWS times-out a query taking too long.

Just for the heck of it, I ran the original query to get the originally reported DNS error (I'm trying to capture logs from the DNS server to get more insight) and my "1000 queries" error is not occurring.  The output is `Error: ThrottlingException: Rate exceeded` <bottom half of screenshot> which is probably from AWS.  113mb.  <yikes!>

![image](https://user-images.githubusercontent.com/35462152/179849048-891c5b67-ea1a-4cfc-909b-1fa4e5934a6d.png)

I feel there's a variable I'm not aware of that was happening Friday but not today.  I did remove my entry from the hosts file and flushed my DNS before running the steampipe queries in the below screenshot so I'm surprised the originally reported behavior is not happening anymore.

With that said, it's more likely that the source of the original issue IS on my side, not something related to steampipe... i just don't know what it can be at the moment.  Unless I can reproduce it again (and present better, more detailed logs) I don't want to spend your time on something that is plausibly a false alarm.

I would say let's close this issue for now.  If i can reproduce it again we can examine it but the last thing I want is involve you folks to chase a ghost.  Sorry for the runaround.

Did want to take a moment to add Steampipe has become an everyday tool for me.  I appreciate what you guys are putting out there!

**Comment by cbruno10** [comment]
@jhodal If you're running into throttling exceptions, what you may find helpful is tuning the config args around error retries, in particular `max_error_retry_attempts` and `min_error_retry_delay`. So for instance, you can try increase both of those so we'll try more before giving up when encountering retryable errors (like throttling), and we'll start with a higher initial backoff period. For more information, please see [AWS Plugin Configuration](https://hub.steampipe.io/plugins/turbot/aws#configuration).

I'll close this issue for now, but if you're still running into DNS or heavy throttling issue, feel free to re-open or open a new issue and we'll be happy to take a look.

Also, glad to see Steampipe has proven useful for you and thanks for the kind words!
